src/vs/base/browser/performance.ts

```diff
@@ -0,0 +1,231 @@
+		queueMicrotask(() => markKeydownEnd());
+	}
+
+	/**
+	 * Mark the end of the keydown event.
+	 */
+	function markKeydownEnd() {
+		// Only measure the first render after keyboard input
+		performance.mark('keydown/end');
+		state.keydown = EventPhase.Finished;
+	}
+
+	/**
+	 * Mark the start of the input event.
+	 */
+	export function markInputStart() {
+		performance.mark('input/start');
+		state.input = EventPhase.InProgress;
+	}
+
+	/**
+	 * Mark the end of the input event.
+	 */
+	export function markInputEnd() {
+		queueMicrotask(() => {
+			performance.mark('input/end');
+			state.input = EventPhase.Finished;
+		});
+	}
+
+	/**
+	 * Mark the start of the animation frame performing the rendering.
+	 */
+	export function markRenderStart() {
+		// Render may be triggered during input, but we only measure the following animation frame
+		if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.render === EventPhase.Before) {
+			// Only measure the first render after keyboard input
+			performance.mark('render/start');
+			state.render = EventPhase.InProgress;
+			queueMicrotask(() => markRenderEnd());
+		}
+	}
+
+	/**
+	 * Mark the end of the animation frame performing the rendering.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	function markRenderEnd() {
+		performance.mark('render/end');
+		state.render = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Mark when the editor textarea selection change event occurs.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	export function markTextareaSelection() {
+		state.selection = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Record the input latency sample if it's ready.
+	 */
+	function record() {
+		// Skip recording this frame if the buffer is full
+		if (measurementsCount >= Constants.bufferLength) {
+			return;
+		}
+		// Selection and render must have finished to record
+		if (state.selection !== EventPhase.Finished || state.render !== EventPhase.Finished) {
+			return;
+		}
+		// Finish the recording, use a timer to ensure that layout/paint is captured. setImmediate
+		// is used if available (Electron) to get slightly more accurate results
+		('setImmediate' in window ? (window as any).setImmediate : setTimeout)(() => {
+			if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.selection === EventPhase.Finished && state.render === EventPhase.Finished) {
+				performance.mark('inputlatency/end');
+
+				performance.measure('keydown', 'keydown/start', 'keydown/end');
+				performance.measure('input', 'input/start', 'input/end');
+				performance.measure('render', 'render/start', 'render/end');
+				performance.measure('inputlatency', 'inputlatency/start', 'inputlatency/end');
+
```

Comment:
What is the advantage with using `performance` instead of using fields/properties where we store `performance.now()`. I imagine that the implementation of `performance.measure(str1, str2, str3)` needs to loop over previous entries and do string comparison, etc. If we'd have simple fields like `this.timeKeydownStart`, `this.timeKeydownEnd` and we'd assign `performance.now()`, we could implement the same in a faster manner?

Reply:
So I'd rather stay on performance mark/measure unless there's some proof it would negatively impact things. Here's some measurements that were done a recently that shows a `performance.mark` call is in the <= 1µs range: https://gist.github.com/paulirish/2fad3834e2617fb199bc12e17058dde4

Reply:
:+1: I like that they show up in dev tools!